### PR TITLE
Make HttpSys client certificate behavior consistent 

### LINKED ISSUE #33586: HttpSys ClientCertificate property renegotiates
ITlsConnectionFeature.ClientCertificate is used to get the client certificate for the current connection.  ITlsConnectionFeature.GetClientCertificateAsync is used to renegotiate the TLS session to request a certificate if you don't already have one. HttpSys has had a behavior in the past where it triggered the renegotiate even from the ClientCertificate property. ClientCertificateMethod.AllowRenegotation was added to control that, but it also disables it for GetClientCertificateAsync.

When [implementing](https://github.com/dotnet/aspnetcore/pull/33264) GetClientCertificateAsync renegotiation for Kestrel it became clear that the pattern developers want is for ClientCertificate to return the current certificate, if any, and for GetClientCertificateAsync to renegotiate for a cert if enabled. This allows them to do conditional logic like buffer the request body before starting the renegotiation.

Proposal: Remove the renegotiate logic from the ClientCertificate property and clean up the SetInitialized logic so that GetClientCertificateAsync can still renegotiate after ClientCertificate is called.
https://github.com/dotnet/aspnetcore/blob/4f2214ab6d614a6f481ce6ca4bcd14a100dd93c2/src/Servers/HttpSys/src/RequestProcessing/RequestContext.FeatureCollection.cs#L325-L328

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes #33586 Now that we know more about the expected usage of the client cert APIs from our work on kestrel, this change brings HttpSys into alignment. The biggest change is that HttpSys' ClientCertificate property used to trigger a sync-over-async renegotiation. In a prior release we made renegotiation opt-in for both the property and the GetClientCertificateAsync method, and this change now removes it entirely for the ClientCertificate property. Anyone that wants renegotiation must opt in and use the GetClientCertificateAsync API. This enables the following usage pattern:
```
if (feature.ClientCertificate == null)
{
  await BufferRequestBodyAsync();
  await feature.GetClientCertificateAsync();
}
```

I also borrowed the test certificates from Kestrel so we could enable these tests on the CI.

### CODE REVIEW

File: src/Servers/HttpSys/src/RequestProcessing/RequestContext.FeatureCollection.cs [hunk 1]
```diff
@@ -335,29 +331,34 @@ string IHttpConnectionFeature.ConnectionId
...
+            }
 
-                if (_clientCert == null && method == ClientCertificateMethod.AllowRenegotation)
-                {
-                    _clientCert = await Request.GetClientCertificateAsync(cancellationToken);
-                }
+            var tlsFeature = (ITlsConnectionFeature)this;
+            var clientCert = tlsFeature.ClientCertificate; // Lazy initialized
+            if (clientCert != null
+                || Server.Options.ClientCertificateMethod != ClientCertificateMethod.AllowRenegotation
```
Comment (BrennanConroy) [inline src/Servers/HttpSys/src/RequestProcessing/RequestContext.FeatureCollection.cs:349]: ClientCertificateMethod.AllowRenegotation == ClientCertificateMode.DelayCertificate?

File: src/Servers/HttpSys/src/RequestProcessing/RequestContext.FeatureCollection.cs [hunk 2]
```diff
@@ -335,29 +331,34 @@ string IHttpConnectionFeature.ConnectionId
             set
             {
                 _clientCert = value;
+                _clientCertTask = Task.FromResult(value);
```
Comment (BrennanConroy) [inline src/Servers/HttpSys/src/RequestProcessing/RequestContext.FeatureCollection.cs:334]: Do we want to do this in Kestrel too?

**Review by BrennanConroy — APPROVED** [review]
>In a prior release we made renegotiation opt-in for both the property and the GetClientCertificateAsync method, and this change now removes it entirely for the ClientCertificate property. Anyone that wants renegotiation must opt in and use the GetClientCertificateAsync API.

Does this need an announcement?

File: src/Servers/HttpSys/src/RequestProcessing/RequestContext.FeatureCollection.cs (hunk 1 above)
Comment (Tratcher) [reply src/Servers/HttpSys/src/RequestProcessing/RequestContext.FeatureCollection.cs:349]: Yes, Kestrel and HttpSys use separate enums. We didn't re-use the old name after learning the mechanics have changed in TLS 1.3 to get rid of renegotiation and add an alternate feature just for client certs.

**Comment by Tratcher** [comment]
Announcement draft:

## ClientCertificate property no longer triggers renegotiation for HttpSys

The `HttpContext.Connection.ClientCertificate` property will no longer trigger TLS renegotiations for HttpSys.

### Version introduced

6.0

### Old behavior

Setting HttpSysOptions.ClientCertificateMethod = ClientCertificateMethod.AllowRenegotation allowed renegotiation to be triggered by both HttpContext.Connection.ClientCertificate and HttpContext.Connection.GetClientCertifiateAsync.

See https://github.com/aspnet/Announcements/issues/422 for related changes in 5.0.

### New behavior

Setting HttpSysOptions.ClientCertificateMethod = ClientCertificateMethod.AllowRenegotation will allow renegotiation to be triggered only by HttpContext.Connection.GetClientCertifiateAsync. HttpContext.Connection.ClientCertificate will return the current certificate if available, but will not renegotiate with the client to request one.

### Reason for change

When implementing the same features for Kestrel it became clear that applications needed to be able to check the state of the client certificate before triggering a renegotiation. This enables the following usage pattern to deal with issues like the request body conflicting with the renegotiation:

```
if (connection.ClientCertificate == null)
{
  await BufferRequestBodyAsync();
  await connection.GetClientCertificateAsync();
}
```

### Recommended action

Applications that use delayed client certificate negotiation need to call GetClientCertificateAsync() to trigger that.

### Category

ASP.NET

### Affected APIs

 [HttpSysOptions.ClientCertificateMethod](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.httpsys.httpsysoptions.clientcertificatemethod)
  [HttpContext.Connection.ClientCertificate](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.connectioninfo.clientcertificate)
  [HttpContext.Connection.GetClientCertifiateAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.connectioninfo.getclientcertificateasync)



---
#### Issue metadata

* Issue type: breaking-change

**Comment by Tratcher** [comment]
https://github.com/aspnet/Announcements/issues/466

**Pull request merged**